RFC-0001-torch-function-for-methods.md

```diff
@@ -0,0 +1,358 @@
+
+Since subclasses are checked before superclasses in `__torch_function__`, it is
+guaranteed that the subclass implementation will be called first. In this
+instance, since `type(self)` is a subclass of all types, the code will
+continue. Since `type(self)` is not `torch.Tensor`, a view into the original
+data is created and returned.
+
+This also works for all operators: `__add__`, `__getitem__` and so on since in
+Python these operators are just dunder methods of the corresponding class.
+
+### The need for `super().__torch_function__`
+To access super, one would do the following:
+```python
+class SubTensor(torch.Tensor):
+    def __torch_function__(self, func, types, args, kwargs):
+        # Pre-processing here
+        val = super().__torch_function__(
+            func,
+            tuple(
+                t for t in types if not issubclass(t, SubTensor)
+            ),
+            args,
+            kwargs
+        )
+        # Post processing here
+```
+
+This way `__torch_function__` knows the list of types to dispatch to, and it
+will _not_ dispatch to `SubTensor` anymore in this example.
+
+To make the need for `super()` to be available concrete, let's consider the
+following scenario:
+
+```python
+class SubTensor(torch.Tensor):
+    def __torch_function__(...):
+        # Pre-processing
+        ret = super().__torch_function__(
+            func,
+            tuple(t if not issubclass(t, SubTensor) else torch.Tensor for t in types),
+            args,
+            kwargs
+        )
+        # Post processing
+        return ret
+
+class SubSubTensor(SubTensor):
+    @torch_function_dispatch(...)
+    def __add__(self, other):
+        # Pre-processing
+        ret = super().__add__()
+        # Post-processing
+        return ret
+```
+
+If `super().__torch_function__` wasn't possible here (or if the signature
+didn't include `types`), then there would be an infinite recursion in this
+instance. This is because `__add__` would call `SubTensor.__torch_function__`,
+`super().__torch_function__` would call `SubSubTensor.__add__` (realise, this
+is a different object than `Tensor.__add__`), which, not realising that it has
+been processed already, would pass control *back* to
+`SubTensor.__torch_function__` and so on ad infinitum.
+
+In this instance, with the proposed changes, processing would follow the
+`__torch_function__` protocol. This means that control would end up in
+`SubTensor.__torch_function__`, go to `Tensor.__torch_function__` from there
+and and then come to `SubSubTensor.__add__`, from where it would go to
+`Tensor.__add__`, and then back up the stack in the reverse order. This means
+that great care needs to be taken when writing `SubTensor.__torch_function__`
+to take into account the fact that it has to handle subclass methods.
+
+### Protocol support for external libraries
+We will also recommend that all `Tensor` subclasses make their own methods go
```

Comment:
It's worth clarifying that this describes *completely new methods* which aren't present on `Tensor`; the reasoning here is equivalent to how we can annotate Python level functions with the dispatch decorator and trigger `__torch_function__` handling (we got rid of the decorator to speed up core code, but for end users this is perfectly fine.)